File: src/test/resources/de/hhu/stups/codegenerator/project7/BString.cpp
#include <iostream>
#include <string>
#include "BString.h"

BString::BString(std::string val) {
    value = val;
}

BString::BString(){};

BString::BString(const BString& val) {
    value = val.value;
}

BBoolean BString::equal(const BString& other) const {
    return value == other.value;
}

BBoolean BString::unequal(const BString& other) const {
    return value != other.value;
}

bool BString::operator <(const BString& other) {
    return value < other.value;
}

void BString::operator =(const BString& other) {
    value = other.value;
}

int BString::hashCode() const {
    return std::hash<std::string>()(value);
}

BBoolean BString::isString() {
    return BBoolean(true);
}

BBoolean BString::isNotString() {
    return BBoolean(false);
}

File: btypes_primitives/src/main/rust/btypes/src/bstring.rs
use std::fmt::{Display, Formatter};
use crate::bobject::BObject;
use crate::bboolean::BBoolean;

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct BString {
    val: String,
}

pub trait TBString {}

impl TBString for BString {}

impl Display for BString {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.val)
    }
}

impl BObject for BString {}

impl BString {
    #![allow(non_snake_case, dead_code)]

    pub fn new(init: &str) -> BString {
        return BString { val: String::from(init) }
    }

    pub fn isString(&self) -> BBoolean {
        return true;
    }

    pub fn isNotString(&self) -> BBoolean {
        return false;
    }

    pub fn equal(&self, other: &BString) -> BBoolean {
        return self.eq(other);
    }

    pub fn unequal(&self, other: &BString) -> BBoolean {
        return !self.eq(other);
    }
}
